### ISSUE TITLE
Many-to-Many Population Suddenly Broken

### ISSUE THREAD

**Issue by macfeaster** [balderdashy/sails#4111]
**Sails version**: 0.12.3
**Node version**: 6.10.3
**NPM version**: 3.10.10
**Operating system**: Linux (Heroku Node.js Buildpack)

<hr/>

I'm using Sails 0.12.3 with PostgreSQL together with the sails-postgresql (0.11.4) module. I have a model, `Mentor` with an association:

```
properties: {
            collection: 'property',
            via: 'mentors',
            through: 'mentor_properties'
}
```

This association has custom column names using a join table definition:

```
identity: 'mentor_properties',

    attributes: {

        property: {
            model: 'property'
        },

        mentor: {
            model: 'mentor'
        }

    }
```

When using the above models in a controller, I have simply called `Mentor.findOne(id).populate('properties')` which has worked flawlessly until within the latest week. I have not updated the code, I have not updated the database - everything is exactly the same, except for a rebuild (which triggers npm install). All dependencies are for exact versions, so no new minor versions with bugs there. The data in the database is untouched.

Populate *seemingly* works as usual, it makes `mentor.properties` an array with `add()` and `remove()`, but the array is empty. I have one-to-many associations on the same model which do work. This only seems to affect many-to-many associations with custom column names.

This behavior is the same on Windows (Node 7.7.4, 4.1.2) as it is on Node 6.10.3 (LTS) with NPM 3 on Heroku. **The exact same code, database settings, dependency versions in package.json and data in the database that worked a week ago, does now not populate properly.** The only logical conclusion I can come up with is that somewhere down in the dependency tree, something broke and it makes many-to-many populations with custom column names silently fail. I do not know what more information to provide, but am happy to try anything upon request. The population works again after I made a Heroku rollback to the old build.

**Comment by sailsbot** [comment]
Hi @macfeaster!  It looks like you missed a step or two when you created your issue.  Please edit your comment (use the pencil icon at the top-right corner of the comment box) and fix the following:

* Verify "I am experiencing a concrete technical issue (aka a bug) with Sails (ideas and feature proposals should follow the guide for proposing features and enhancements (http://bit.ly/sails-feature-guide), which involves making a pull request).  If you're not 100% certain whether it's a bug or not, that's okay--you may continue.  The worst that can happen is that the issue will be closed and we'll point you in the right direction."
* Verify "I am not asking a question about how to use Sails or about whether or not Sails has a certain feature (please refer to the documentation(http://sailsjs.com), or post on http://stackoverflow.com, our Google Group (http://bit.ly/sails-google-group) or our live chat (https://gitter.im/balderdashy/sails)."
* Verify "I have already searched for related issues, and found none open (if you found a related _closed_ issue, please link to it in your post)."
* Verify "My issue title is concise, on-topic and polite ("jst.js being removed from layout.ejs on lift" is good; "templates dont work" or "why is sails dumb" are not so good)."
* Verify "I have tried all the following (if relevant) and my issue remains:"
* Verify "**I can provide steps to reproduce this issue that others can follow.**"

As soon as those items are rectified, post a new comment (e.g. &ldquo;Ok, fixed!&rdquo;) below and we'll take a look.  Thanks!

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@sailsjs.com

**Comment by macfeaster** [comment]
Ok, fixed!

**Comment by sailsbot** [comment]
Sorry to be a hassle, but it looks like your issue is still missing some required info.  Please double-check your initial comment and try again.

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@sailsjs.com

**Comment by sailsbot** [comment]
Sorry to be a hassle, but it looks like your issue is still missing some required info.  Please double-check your initial comment and try again.

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@sailsjs.com

**Comment by macfeaster** [comment]
Ok, fixed!

**Comment by sailsbot** [comment]
@macfeaster Thanks for posting, we'll take a look as soon as possible.

---

For help with questions about Sails, [click here](http://sailsjs.com/support).  If you&rsquo;re interested in hiring @sailsbot and her minions in Austin, [click here](http://sailsjs.com/studio).

**Comment by sgress454** [comment]
@macfeaster If you want to follow up on your hunch about dependencies, you could try logging into one of the Heroku boxes and running `npm ls` to get the whole dependency tree, and then doing the same with a new build that reproduces the problem.  If you can try that and see if there are any discrepancies, that might help to pinpoint what changed.

**Comment by macfeaster** [comment]
Thank you for the help. I did `npm ls` on the working build and a fresh build. There are some sub-dependencies of the `pg` dependency of `connect-pg-simple` which have changed, but shouldn't they relate to sessions and have no effect on Waterline? Otherwise, everything is the same. I'm so incredibly confused by this bug since the `through` association just stopped working for seemingly no reason.

As a quick fix, I tried to use the through model directly, but it is undefined even though the sailsjs documentation clearly states that it can be queried directly. How can I query the join table directly?
